### ISSUE TITLE
fix: `Unsupported class file major version 65` when running `shorebird init` on Linux

### ISSUE THREAD

**Issue by bryanoltman** [shorebirdtech/shorebird#1821]
Reported on Discord ([chat link](https://discord.com/channels/1030243211995791380/1221828343746531358/1221832057118920906))

Output:

```
shorebird init -v                                                                                                         ✔ 
[Process.run] /home/kingwill101/code/linkboard/frontend/android/gradlew app:tasks --all --console=auto (in /home/kingwill101/code/linkboard/frontend/android)
⠼ Detecting product flavors... (3.5s)Exited with code 1

stdout:
> Task :gradle:compileJava NO-SOURCE
> Task :gradle:compileGroovy FAILED
1 actionable task: 1 executed


stderr:

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':gradle:compileGroovy'.
> BUG! exception in phase 'semantic analysis' in source unit '/home/kingwill101/flutter/packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy' Unsupported class file major version 65

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 3s

✗ Detecting product flavors (3.6s)
Unable to extract product flavors.
Exception: > Task :gradle:compileJava NO-SOURCE
> Task :gradle:compileGroovy FAILED
1 actionable task: 1 executed


FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':gradle:compileGroovy'.
> BUG! exception in phase 'semantic analysis' in source unit '/home/kingwill101/flutter/packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy' Unsupported class file major version 65

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 3s
```

**Comment by eseidel** [comment]
I presume this is a java version problem?  e.g. that whatever version of `java` we're finding on disk is not the one Gradle wanted?

**Comment by kingwill101** [comment]
> I presume this is a java version problem? e.g. that whatever version of `java` we're finding on disk is not the one Gradle wanted?

shorebird is not using the version bundled with android studio. flutter however finds and uses the correct version.

```
flutter doctor -v                                                                                                                                                              ✔ 
[✓] Flutter (Channel stable, 3.19.4, on Manjaro Linux 6.6.19-1-MANJARO, locale en_US.UTF-8)
    • Flutter version 3.19.4 on channel stable at /home/kingwill101/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 68bfaea224 (5 days ago), 2024-03-20 15:36:31 -0700
    • Engine revision a5c24f538d
    • Dart version 3.3.2
    • DevTools version 2.31.1

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /home/kingwill101/Android/Sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /home/kingwill101/.local/share/JetBrains/Toolbox/apps/android-studio/jbr/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = /usr/bin/google-chrome-stable

[✓] Linux toolchain - develop for Linux desktop
    • clang version 16.0.6
    • cmake version 3.28.3
    • ninja version 1.11.1
    • pkg-config version 2.1.0

[✓] Android Studio (version 2023.2)
    • Android Studio at /home/kingwill101/.local/share/JetBrains/Toolbox/apps/android-studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • android-studio-dir = /home/kingwill101/.local/share/JetBrains/Toolbox/apps/android-studio
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.3)
    • IntelliJ at /home/kingwill101/.local/share/JetBrains/Toolbox/apps/intellij-idea-ultimate-2
    • Flutter plugin version 78.4.2
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] IntelliJ IDEA Ultimate Edition (version 2023.2)
    • IntelliJ at /home/kingwill101/.local/share/JetBrains/Toolbox/apps/intellij-idea-ultimate
    • Flutter plugin version 76.3.4
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] Connected device (2 available)
    • Linux (desktop) • linux  • linux-x64      • Manjaro Linux 6.6.19-1-MANJARO
    • Chrome (web)    • chrome • web-javascript • Google Chrome 123.0.6312.58

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

shorebird doctor -v                                                                                                                                                        INT ✘ 
Shorebird 0.27.2 • git@github.com:shorebirdtech/shorebird.git
Flutter 3.19.4 • revision 6e1c9add71baf57b4178d4d315cea7ba52afe355
Engine • revision 8a7314ec11ad0e9b5b801fc6ed50cc70a873f157

Android Toolchain
  • Android Studio: not detected
  • Android SDK: /home/kingwill101/Android/Sdk
  • ADB: /home/kingwill101/Android/Sdk/platform-tools/adb
  • JAVA_HOME: not detected

✓ Shorebird is up-to-date (1.3s)
✓ Flutter install is correct (0.2s)
✓ Has access to storage.googleapis.com (0.4s)

No issues detected!

to work around the issue I had to manually set JAVA_HOME to the bundled version that comes with android studio

**Comment by eseidel** [comment]
Yeah, we've had several rounds of futzing with our Java detection (e.g. https://github.com/shorebirdtech/shorebird/issues/1163).  Sounds like we need one more round.

**Comment by bryanoltman** [comment]
From my Ubuntu laptop:

```
➜  linux_test shorebird doctor -v
Shorebird 0.27.3 • git@github.com:shorebirdtech/shorebird.git
Flutter 3.19.4 • revision 6e1c9add71baf57b4178d4d315cea7ba52afe355
Engine • revision 8a7314ec11ad0e9b5b801fc6ed50cc70a873f157

Android Toolchain
  • Android Studio: not detected
  • Android SDK: /home/bryan/Android/Sdk
  • ADB: /home/bryan/Android/Sdk/platform-tools/adb
  • JAVA_HOME: not detected

✓ Shorebird is up-to-date (1.1s)
✓ Flutter install is correct (0.4s)
✓ AndroidManifest.xml files contain INTERNET permission (73ms)
✓ Has access to storage.googleapis.com (0.2s)

No issues detected!
```

```
➜  linux_test flutter doctor -v
[✓] Flutter (Channel stable, 3.19.4, on Ubuntu 22.04.4 LTS 6.5.0-26-generic,
    locale en_US.UTF-8)
    • Flutter version 3.19.4 on channel stable at /home/bryan/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 68bfaea224 (6 days ago), 2024-03-20 15:36:31 -0700
    • Engine revision a5c24f538d
    • Dart version 3.3.2
    • DevTools version 2.31.1

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /home/bryan/Android/Sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /snap/android-studio/151/jbr/bin/java
    • Java version OpenJDK Runtime Environment (build
      17.0.9+0-17.0.9b1087.7-11185874)
    • All Android licenses accepted.

[✗] Chrome - develop for the web (Cannot find Chrome executable at
    google-chrome)
    ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.

[✗] Linux toolchain - develop for Linux desktop
    • Ubuntu clang version 14.0.0-1ubuntu1.1
    • cmake version 3.22.1
    • ninja version 1.10.1
    • pkg-config version 0.29.2
    ✗ GTK 3.0 development libraries are required for Linux development.
      They are likely available from your distribution (e.g.: apt install
      libgtk-3-dev)

[✓] Android Studio (version 2023.2)
    • Android Studio at /snap/android-studio/151
    • Flutter plugin version 78.4.1
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      17.0.9+0-17.0.9b1087.7-11185874)

[✓] VS Code (version 1.87.2)
    • VS Code at /snap/code/current/usr/share/code
    • Flutter extension version 3.70.0

[✓] Connected device (1 available)
    • Linux (desktop) • linux • linux-x64 • Ubuntu 22.04.4 LTS 6.5.0-26-generic

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 2 categories.
```

I am able to successfully run `shorebird init` and `shorebird release android`, so I don't think the issue is Java detection, but possibly Java version. I see the following when running `java --version`:

```
openjdk 11.0.22 2024-01-16
OpenJDK Runtime Environment (build 11.0.22+7-post-Ubuntu-0ubuntu222.04.1)
OpenJDK 64-Bit Server VM (build 11.0.22+7-post-Ubuntu-0ubuntu222.04.1, mixed mode, sharing)
```

**Comment by bryanoltman** [comment]
So this is perplexing. We have the same Java and Android Studio versions, yet this is working for me. @kingwill101, what version is the `gradlew` that's used when you run `shorebird init`? Here's what I see:

```
➜  linux_test ./android/gradlew --version

------------------------------------------------------------
Gradle 7.6.3
------------------------------------------------------------

Build time:   2023-10-04 15:59:47 UTC
Revision:     1694251d59e0d4752d547e1fd5b5020b798a7e71

Kotlin:       1.7.10
Groovy:       3.0.13
Ant:          Apache Ant(TM) version 1.10.11 compiled on July 10 2021
JVM:          11.0.22 (Ubuntu 11.0.22+7-post-Ubuntu-0ubuntu222.04.1)
OS:           Linux 6.5.0-26-generic amd64
```

**Comment by kingwill101** [comment]
> So this is perplexing. We have the same Java and Android Studio versions, yet this is working for me. @kingwill101, what version is the `gradlew` that's used when you run `shorebird init`? Here's what I see:
> 
> ```
> ➜  linux_test ./android/gradlew --version
> 
> ------------------------------------------------------------
> Gradle 7.6.3
> ------------------------------------------------------------
> 
> Build time:   2023-10-04 15:59:47 UTC
> Revision:     1694251d59e0d4752d547e1fd5b5020b798a7e71
> 
> Kotlin:       1.7.10
> Groovy:       3.0.13
> Ant:          Apache Ant(TM) version 1.10.11 compiled on July 10 2021
> JVM:          11.0.22 (Ubuntu 11.0.22+7-post-Ubuntu-0ubuntu222.04.1)
> OS:           Linux 6.5.0-26-generic amd64
> ```
```
./gradlew --version                    ✔ 

------------------------------------------------------------
Gradle 7.6.3
------------------------------------------------------------

Build time:   2023-10-04 15:59:47 UTC
Revision:     1694251d59e0d4752d547e1fd5b5020b798a7e71

Kotlin:       1.7.10
Groovy:       3.0.13
Ant:          Apache Ant(TM) version 1.10.11 compiled on July 10 2021
JVM:          21.0.2 (Oracle Corporation 21.0.2+13)
OS:           Linux 6.6.19-1-MANJARO amd64

```

**Comment by wycers** [comment]
I solve this issue by upgrading the gradle. See [this post](https://stackoverflow.com/questions/75155059/flutter-unsupported-class-file-major-version-63-android-studio).

**Comment by eseidel** [comment]
The fix here is that `shorebird doctor` should verify specific versions of java and gradle are available.

**Comment by erickzanardo** [comment]
@eseidel this was kinda done by https://github.com/shorebirdtech/shorebird/pull/2174

Gradle we can't really show much on the doctor, since flutter uses the gradlew, which is "embeded" in the project.

**Comment by eseidel** [comment]
Does `gradlew` have a `--version` command we can check?  There are per-project checks which we do separately from per-system checks.

**Comment by erickzanardo** [comment]
> Does `gradlew` have a `--version` command we can check? There are per-project checks which we do separately from per-system checks.

Oh, interesting, yes, `gradlew` have a `--version`, it is actually not much different from a gradle installed globally, just that it is "embed" in the project.

Here is an example:

```
./gradlew --version

------------------------------------------------------------
Gradle 7.6.3
------------------------------------------------------------

Build time:   2023-10-04 15:59:47 UTC
Revision:     1694251d59e0d4752d547e1fd5b5020b798a7e71

Kotlin:       1.7.10
Groovy:       3.0.13
Ant:          Apache Ant(TM) version 1.10.11 compiled on July 10 2021
JVM:          11.0.23 (Azul Systems, Inc. 11.0.23+9-LTS)
OS:           Mac OS X 14.4.1 aarch64
```

Should we add that info to the doctor when the user is in a shorebird project?

**Comment by eseidel** [comment]
> Should we add that info to the doctor when the user is in a shorebird project?

I don't know. 🤷  My guess is that the java version is causing this specific issue (probably not the gradle version).

**Comment by erickzanardo** [comment]
I think the java version is the issue too!

**Comment by erickzanardo** [comment]
Got some more interesting information from Discord, an user reported having a similar issue, but on MacOS, and the error was the same, but with version 66 instead of 65.

<https://discord.com/channels/1030243211995791380/1252240492696178698>

An interesting aspect of the report is that the user doesn't have AS installed, so their java path comes directly from an homebrew installation:

```
Android Toolchain
  • Android Studio: not detected
  • Android SDK: not detected
  • ADB: not detected
  • JAVA_HOME: not detected
  • JAVA_EXECUTABLE: /usr/bin/java
  • JAVA_VERSION: openjdk version "22.0.1" 2024-04-16
                  OpenJDK Runtime Environment Homebrew (build 22.0.1)
                  OpenJDK 64-Bit Server VM Homebrew (build 22.0.1, mixed mode, 
```

After some research I found a Java/Gradle compability Matrix: <https://docs.gradle.org/current/userguide/compatibility.html>

And in it, it states that the user's java version 22, is not compatible with the java version.

So I believe that the solution for this is to identify the gradle version, the java one, check if they are compatible, and fail with a more detailed message in case they are not.

**Comment by eseidel** [comment]
This feels like something `flutter doctor` should eventually cover, but we can start by doing gradle/java checks in `shorebird doctor` and later port it to `flutter` if useful.

**Comment by erickzanardo** [comment]
This was not supposed to be closed by #2255, there is still some changes to land before considering this closed, reopening
